calendar/api/parent/ext-calendar-provider.js

```diff
@@ -103,11 +103,52 @@ class ExtCalendarProvider extends cal.provider.BaseClass {
           : ["unsubscribe"];
       case "requiresNetwork":
         return !(this.capabilities.requires_network === false);
+      case "itip.transport":
+        if (this.extension.emitter.has("calendar.provider.onSend")) {
+          return this;
+        }
+        break;
     }
 
     return super.getProperty(name);
   }
 
+  scheme = "mailto";
+
+  async sendItems(aRecipients, aItipItem) {
+    let { jsmime } = ChromeUtils.import("resource:///modules/jsmime.jsm");
+    let method = aItipItem.responseMethod;
+    let transport = super.getProperty("itip.transport").wrappedJSObject;
+    let { subject, body } = transport._prepareItems(aItipItem);
+    let serializer = Cc["@mozilla.org/calendar/ics-serializer;1"].createInstance(Ci.calIIcsSerializer);
+    let itemList = aItipItem.getItemList();
+    serializer.addItems(itemList, itemList.length);
+    let methodProp = cal.getIcsService().createIcalProperty("METHOD");
+    methodProp.value = method;
+    serializer.addProperty(methodProp);
+    let icsText = serializer.serializeToString();
+    let randomGenerator = Cc["@mozilla.org/security/random-generator;1"].getService(Ci.nsIRandomGenerator);
+    let randomBytes = randomGenerator.generateRandomBytes(12);
+    let boundary = randomBytes.map(byte => byte.toString(16).padStart(2, 0)).join("").toUpperCase();
```

Comment:
Doesn't JSMime have functions to generate a multipart MIME structure with boundaries, instead of sticking strings and newlines together?

Reply:
No, that's still handled by nsMsgSendPart.cpp for now.

Reply:
I think js message sending is available behind a pref, might be worth looking in to.